src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs

```diff
@@ -0,0 +1,189 @@
+					break;
+
+				// this is the case when the left hand side of the
+				// lambda has parens
+				// eg () => 1, (x, y) => x + y
+				case SyntaxKind.ParenthesizedLambdaExpression:
+					if( !HasCaptures( context, argument ) ) {
+						return;
+					}
+					break;
+
+				// this is the case when the left hand side of the
+				// lambda does not have parens
+				// eg x => x + 1
+				case SyntaxKind.SimpleLambdaExpression:
+					if( !HasCaptures( context, argument ) ) {
+						return;
+					}
+					break;
+
+				// this is the case where an expression is invoked,
+				// which returns a Func<T>
+				// eg ( () => { return () => 1 } )()
+				case SyntaxKind.InvocationExpression:
+					// we are rejecting this because it is tricky to
+					// analyze properly, but also a bit ugly and should
+					// never really be necessary
+					break;
+
+				default:
+					// we need StatelessFunc<T> to be ultra safe, so we'll
+					// reject usages we do not understand yet
+					break;
+			}
+
+			var diag = Diagnostic.Create(
+				Diagnostics.StatelessFuncIsnt,
+				argument.GetLocation()
+			);
+
+			context.ReportDiagnostic( diag );
+		}
+
+		private static bool IsStaticMemberAccess(
+			SyntaxNodeAnalysisContext context,
+			ExpressionSyntax expression
+		) {
+			ISymbol symbol = context
+				.SemanticModel
+				.GetSymbolInfo( expression ).Symbol;
+
+			if( symbol == null ) {
+				return false;
+			}
+
+			return symbol.IsStatic;
+		}
+
+		private static bool HasCaptures(
+			SyntaxNodeAnalysisContext context,
+			ExpressionSyntax expression
+		) {
+
+			DataFlowAnalysis dataFlow = context
+				.SemanticModel
+				.AnalyzeDataFlow( expression );
+
+			ImmutableArray<ISymbol> captures = dataFlow.Captured;
+			return captures.Length > 0;
+		}
+
+		private static bool IsStatelessFunc(
+			ISymbol symbol,
+			SyntaxNodeAnalysisContext context
+		) {
+			ImmutableHashSet<ISymbol> statelessFuncs = GetStatelessFuncTypes( context.Compilation );
+
+			if( statelessFuncs.Contains( symbol ) ) {
+				// we've found a definition that matches exactly with the symbol
+				return true;
+			}
+
+			// Generics work a bit different, in that the symbol we have to work
+			// with is not (eg StatelessFunc<int> ), but the list of symbols we're
+			// checking against are the definitions, which are (eg
+			// StatelessFunc<T> ) generic. So check the "parent" definition.
+			return statelessFuncs.Contains( symbol.OriginalDefinition );
+		}
+
+		private static ImmutableHashSet<ISymbol> GetStatelessFuncTypes( Compilation compilation) {
+
+			var builder = ImmutableHashSet.CreateBuilder<ISymbol>();
+
+			var types = new string[] {
+				"D2L.StatelessFunc`1",
+				"D2L.StatelessFunc`2",
+				"D2L.StatelessFunc`3",
+				"D2L.StatelessFunc`4",
+				"D2L.StatelessFunc`5",
+				"D2L.StatelessFunc`6",
+			};
+
+			foreach( string typeName in types ) {
+				INamedTypeSymbol typeSymbol = compilation.GetTypeByMetadataName( typeName );
```

Comment:
It might be nice to cache this `ImmutableHashSet<ISymbol>`. I don't 100% remember what the perf of `GetTypeByMetadataName` looked like (probably pretty fast once caches are primed, at least.)

How I've done this before is something like call `GetStatelessFunctTypes` in `RegisterAnalysis` and then pass it to `AnalyzeInvocation` (that way its still different per compilation, but it gets totally cached.